File: udp-py/udp/hub.py
# -*- coding: utf-8 -*-
#
#   UDP: User Datagram Protocol
#
#                                Written in 2020 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2020 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

import socket
from abc import ABC
from typing import Optional, Iterable

from startrek.types import SocketAddress, AddressPairMap
from startrek.skywalker import Runner
from startrek import Channel, BaseChannel
from startrek import Connection, ConnectionDelegate
from startrek import BaseConnection, ActiveConnection
from startrek import BaseHub

from .channel import PacketChannel


class ChannelPool(AddressPairMap[Channel]):

    # Override
    def get(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        assert not (remote is None and local is None), 'both addresses are empty'
        # -- Step 1 --
        #     (remote, local)
        #         this step will get the channel connected to the remote address
        #         and bound to the local address at the same time;
        #     (null, local)
        #         this step will get a channel bound to the local address, and
        #         surely not connected;
        #     (remote, null)
        #         this step will get a channel connected to the remote address,
        #         no matter which local address to be bound;
        channel = super().get(remote=remote, local=local)
        if channel is not None:
            return channel
        elif remote is None:
            # failed to get a channel bound to the local address
            return None
        # -- Step 2 --
        #     (remote, local)
        #         try to get a channel which bound to the local address, but
        #         not connected to any remote address;
        #     (null, local)
        #         this situation has already done at step 1;
        if local is not None:
            # ignore the remote address
            channel: Channel = super().get(remote=None, local=local)
            if channel is not None and channel.remote_address is None:
                # got a channel not connected yet
                return channel
        # -- Step 3 --
        #     (remote, null)
        #         try to get a channel that bound to any local address, but
        #         not connected yet;
        array = self.items
        for item in array:
            if item.remote_address is None:
                return item
        # not found

    # Override
    def set(self, item: Optional[Channel],
            remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        # 1. remove cached item
        cached = super().remove(item=item, remote=remote, local=local)
        if cached is not None and cached is not item:
            Runner.async_task(coro=cached.close())
        # 2. set new item
        old = super().set(item=item, remote=remote, local=local)
        assert old is None, 'should not happen: %s' % old
        return cached

    # Override
    def remove(self, item: Optional[Channel],
               remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        cached = super().remove(item=item, remote=remote, local=local)
        if cached is not None and cached is not item:
            Runner.async_task(coro=cached.close())
        if item is not None:
            Runner.async_task(coro=item.close())
        return cached


# noinspection PyAbstractClass
class PacketHub(BaseHub, ABC):
    """ Base Datagram Hub """

    def __init__(self, delegate: ConnectionDelegate):
        super().__init__(delegate=delegate)
        self.__channel_pool = self._create_channel_pool()

    # noinspection PyMethodMayBeStatic
    def _create_channel_pool(self):
        return ChannelPool()

    async def bind(self, address: SocketAddress = None, host: str = None, port: int = 0):
        if address is None:
            assert host is not None and port > 0, 'address error: (%s:%d)' % (host, port)
            address = (host, port)
        channel = self.__channel_pool.get(remote=None, local=address)
        if channel is None:
            channel = self._create_channel(remote=None, local=address)
            assert isinstance(channel, BaseChannel), 'channel error: %s, %s' % (address, channel)
            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
            sock.setblocking(True)
            sock.bind(address)
            sock.setblocking(False)
            # set socket for this channel
            await channel.set_socket(sock=sock)
            self.__channel_pool.set(item=channel, remote=None, local=address)

    #
    #   Channel
    #

    # noinspection PyMethodMayBeStatic
    def _create_channel(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Channel:
        # override for user-customized channel
        return PacketChannel(remote=remote, local=local)

    # Override
    def _all_channels(self) -> Iterable[Channel]:
        """ get a copy of all channels """
        return self.__channel_pool.items

    # Override
    def _remove_channel(self, channel: Optional[Channel],
                        remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ remove cached channel """
        return self.__channel_pool.remove(item=channel, remote=remote, local=local)

    def _get_channel(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ get cached channel """
        return self.__channel_pool.get(remote=remote, local=local)

    def _set_channel(self, channel: Channel,
                     remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        """ cache channel """
        return self.__channel_pool.set(item=channel, remote=remote, local=local)


class ServerHub(PacketHub):
    """ Datagram Server Hub """

    # Override
    def _create_connection(self, remote: SocketAddress, local: Optional[SocketAddress]) -> Optional[Connection]:
        conn = BaseConnection(remote=remote, local=local)
        conn.delegate = self.delegate  # gate
        return conn

    # Override
    async def open(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        # get channel with direction (remote, local)
        return self._get_channel(remote=remote, local=local)


class ClientHub(PacketHub):
    """ Datagram Client Hub """

    # Override
    def _create_connection(self, remote: SocketAddress, local: Optional[SocketAddress]) -> Optional[Connection]:
        conn = ActiveConnection(remote=remote, local=local)
        conn.delegate = self.delegate  # gate
        return conn

    # Override
    async def open(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]) -> Optional[Channel]:
        # get channel with direction (remote, local)
        return self._get_channel(remote=remote, local=local)

File: startrek-js/StarTrek/src/socket/hub.js
'use strict';
// license: https://mit-license.org
//
//  Star Trek: Interstellar Transport
//
//                               Written in 2022 by Moky <albert.moky@gmail.com>
//
// =============================================================================
// The MIT License (MIT)
//
// Copyright (c) 2022 Albert Moky
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================
//

//! require 'type/apm.js'
//! require 'net/hub.js'

    st.socket.ConnectionPool = function () {
        AddressPairMap.call(this);
    };
    var ConnectionPool = st.socket.ConnectionPool;

    Class(ConnectionPool, AddressPairMap, null, {

        // Override
        set: function (remote, local, value) {
            // remove cached item
            var cached = AddressPairMap.prototype.remove.call(this, remote, local, value);
            // if (cached && cached !== value) {
            //     cached.close();
            // }
            AddressPairMap.prototype.set.call(this, remote, local, value);
            return cached;
        }

        // // Override
        // remove: function (remote, local, value) {
        //     var cached = AddressPairMap.prototype.remove.call(this, remote, local, value);
        //     if (cached && cached !== value) {
        //         cached.close();
        //     }
        //     if (value) {
        //         value.close();
        //     }
        //     return cached;
        // }
    });

    /**
     *  Base Hub
     *  ~~~~~~~~
     *
     * @param {ConnectionDelegate} gate
     */
    st.socket.BaseHub = function (gate) {
        BaseObject.call(this);
        this.__delegate = gate;
        this.__connPool = this.createConnectionPool();
        // last time drove connections
        this.__last = new Date();
    };
    var BaseHub = st.socket.BaseHub;

    Class(BaseHub, BaseObject, [Hub], null);

    // protected
    BaseHub.prototype.createConnectionPool = function () {
        return new ConnectionPool();
    };

    // protected: delegate for handling connection events
    BaseHub.prototype.getDelegate = function () {
        return this.__delegate;
    };

    /*  Maximum Segment Size
     *  ~~~~~~~~~~~~~~~~~~~~
     *  Buffer size for receiving package
     *
     *  MTU        : 1500 bytes (excludes 14 bytes ethernet header & 4 bytes FCS)
     *  IP header  :   20 bytes
     *  TCP header :   20 bytes
     *  UDP header :    8 bytes
     */
    BaseHub.MSS = 1472;  // 1500 - 20 - 8

    //
    //  Channel
    //

    /**
     *  Get all channels
     *
     * @return {Channel[]} copy of channels
     */
    // protected
    BaseHub.prototype.allChannels = function () {};

    /**
     *  Remove socket channel
     *
     * @param {SocketAddress} remote
     * @param {SocketAddress} local
     * @param {Channel} channel
     */
    // protected
    BaseHub.prototype.removeChannel = function (remote, local, channel) {};

    //
    //  Connections
    //

    /**
     *  Create connection with socket channel & addresses
     *
     * @param {SocketAddress} remote
     * @param {SocketAddress} local
     * @return {Connection}
     */
    // protected
    BaseHub.prototype.createConnection = function (remote, local) {};

    // protected
    BaseHub.prototype.allConnections = function () {
        return this.__connPool.items();
    };

    // protected
    BaseHub.prototype.getConnection = function (remote, local) {
        return this.__connPool.get(remote, local);
    };

    // protected
    BaseHub.prototype.setConnection = function (remote, local, connection) {
        return this.__connPool.set(remote, local, connection);
    };

    // protected
    BaseHub.prototype.removeConnection = function (remote, local, connection) {
        return this.__connPool.remove(remote, local, connection);
    };

    //
    //  Hub
    //

    // Override
    BaseHub.prototype.connect = function (remote, local) {
        //
        //  0. pre-checking
        //
        var conn = this.getConnection(remote, local);
        if (conn) {
            // check local address
            if (!local) {
                return conn;
            }
            var address = conn.getLocalAddress();
            if (!address || address.equals(local)) {
                return conn;
            }
        }
        //
        //  1. create new connection & cache it
        //
        conn = this.createConnection(remote, local);
        if (!local) {
            local = conn.getLocalAddress();
        }
        // cache the connection
        var cached = this.setConnection(remote, local, conn);
        if (cached && cached !== conn) {
            cached.close();
        }
        //
        //  2. start the new connection
        //
        if (conn instanceof BaseConnection) {
            // try to open channel with direction (remote, local)
            conn.start(this);
        }
        return conn;
    };

    //
    //  Processor
    //

    // protected
    BaseHub.prototype.closeChannel = function (channel) {
        try {
            if (channel.isOpen()) {
                channel.close();
            }
        } catch (e) {
            // console.error(this, 'close channel error', e, this);
        }
    };

    // protected
    BaseHub.prototype.driveChannel = function (channel) {
        //
        //  0. check channel state
        //
        var cs = channel.getState();
        if (StateOrder.INIT.equals(cs)) {
            // preparing
            return false;
        } else if (StateOrder.CLOSED.equals(cs)) {
            // finished
            return false;
        }
        // cs == opened
        // cs == alive
        var conn;
        var remote;
        var local;
        var data;           // Uint8Array
        //
        //  1. try to receive
        //
        try {
            var pair = channel.receive(BaseHub.MSS);
            data = pair.a;
            remote = pair.b;
        } catch (e) {
            remote = channel.getRemoteAddress();
            local = channel.getLocalAddress();
            var gate = this.getDelegate();
            var cached;
            if (!gate || !remote) {
                // UDP channel may not connected,
                // so no connection for it
                cached = this.removeChannel(remote, local, channel);
            } else {
                // remove channel and callback with connection
                conn = this.getConnection(remote, local);
                cached = this.removeChannel(remote, local, channel);
                if (conn) {
                    gate.onConnectionError(e, conn);
                }
            }
            if (cached && cached !== channel) {
                this.closeChannel(cached);
            }
            this.closeChannel(channel);
            return false;
        }
        if (!remote || !data/* || data.length === 0*/) {
            // received nothing
            return false;
        } else {
            local = channel.getLocalAddress();
        }
        //
        //  2. get connection for processing received data
        //
        conn = this.connect(remote, local);
        if (conn) {
            conn.onReceivedData(data);
        }
        return true;
    };

    // protected
    BaseHub.prototype.driveChannels = function (channels) {
        var count = 0;
        for (var i = channels.length - 1; i >= 0; --i) {
            // drive channel to receive data
            if (this.driveChannel(channels[i])) {
                ++count;
            }
        }
        return count;
    };

    // protected
    BaseHub.prototype.cleanupChannels = function (channels) {
        var cached, sock;   // Channel
        var remote, local;  // SocketAddress
        for (var i = channels.length - 1; i >= 0; --i) {
            sock = channels[i];
            if (!sock.isOpen()) {
                // if channel not connected (TCP) and not bound (UDP),
                // means it's closed, remove it from the hub
                remote = sock.getRemoteAddress();
                local = sock.getLocalAddress();
                cached = this.removeChannel(remote, local, sock);
                if (cached && cached !== sock) {
                    this.closeChannel(cached);
                }
            }
        }
    };

    // protected
    BaseHub.prototype.driveConnections = function (connections) {
        var now = new Date();
        var elapsed = Duration.between(this.__last, now);
        for (var i = connections.length - 1; i >= 0; --i) {
            // drive connection to go on
            connections[i].tick(now, elapsed);
            // NOTICE: let the delegate to decide whether close an error connection
            //         or just remove it.
        }
        this.__last = now;
    };

    // protected
    BaseHub.prototype.cleanupConnections = function (connections) {
        var cached, conn;   // Connection
        var remote, local;  // SocketAddress
        for (var i = connections.length - 1; i >= 0; --i) {
            conn = connections[i];
            if (!conn.isOpen()) {
                // if connection closed, remove it from the hub; notice that
                // ActiveConnection can reconnect, it'll be not connected
                // but still open, don't remove it in this situation.
                remote = conn.getRemoteAddress();
                local = conn.getLocalAddress();
                cached = this.removeConnection(remote, local, conn);
                if (cached && cached !== conn) {
                    cached.close();
                }
            }
        }
    };

    // Override
    BaseHub.prototype.process = function () {
        // 1. drive all channels to receive data
        var channels = this.allChannels();
        var count = this.driveChannels(channels);
        // 2. drive all connections to move on
        var connections = this.allConnections();
        this.driveConnections(connections);
        // 3. cleanup closed channels and connections
        this.cleanupChannels(channels);
        this.cleanupConnections(connections);
        return count > 0;
    };
